### PR TITLE
Fixed #16281 -- Fixed ContentType.get_object_for_this_type() in a multiple database setup.

### PR DESCRIPTION
**Pull request by bcail**

### CODE REVIEW

File: django/contrib/contenttypes/models.py [hunk 1]
```diff
@@ -174,20 +174,23 @@ def model_class(self):
         except LookupError:
             return None
 
-    def get_object_for_this_type(self, **kwargs):
+    def get_object_for_this_type(self, using=None, **kwargs):
```
Comment (felixxm) [inline django/contrib/contenttypes/models.py:177]: This method is [documented](https://docs.djangoproject.com/en/5.0/ref/contrib/contenttypes/#django.contrib.contenttypes.models.ContentType.get_object_for_this_type), so any change in the signature should be reflected in docs.

File: django/contrib/contenttypes/models.py [hunk 2]
```diff
@@ -174,20 +174,23 @@ def model_class(self):
...
         """
         Return an object of this type for the keyword arguments given.
         Basically, this is a proxy around this object_type's get_object() model
         method. The ObjectNotExist exception, if thrown, will not be caught,
         so code that calls this method should catch it.
         """
-        return self.model_class()._base_manager.using(self._state.db).get(**kwargs)
+        if using:
+            return self.model_class()._base_manager.using(using).get(**kwargs)
```
Comment (felixxm) [inline django/contrib/contenttypes/models.py:185]: This path is not covered by tests. Do we need this change?

**Review by felixxm — COMMENTED** [review]
@bcail Thanks :+1: I left comments.

File: django/contrib/contenttypes/models.py (hunk 2 above)
Comment (bcail) [reply django/contrib/contenttypes/models.py:185]: Thanks, @felixxm. If I remove that path, I get the follow test error:
```
ERROR: test_generic_key_separation (multiple_database.tests.QueryTestCase.test_generic_key_separation)
AttributeError: 'NoneType' object has no attribute 'title'
```

File: django/contrib/contenttypes/models.py (hunk 1 above)
Comment (bcail) [reply django/contrib/contenttypes/models.py:177]: @felixxm I updated the docs.

File: docs/ref/contrib/contenttypes.txt [hunk 3]
```diff
@@ -106,14 +106,17 @@ methods that allow you to get from a
...
+.. method:: ContentType.get_object_for_this_type(using=None, **kwargs)
 
     Takes a set of valid :ref:`lookup arguments <field-lookups-intro>` for the
     model the :class:`~django.contrib.contenttypes.models.ContentType`
     represents, and does
     :meth:`a get() lookup <django.db.models.query.QuerySet.get>`
     on that model, returning the corresponding object.
 
+    The query uses the default database, unless the ``using`` argument
+    specifies a different database.
```
Comment (felixxm) [inline docs/ref/contrib/contenttypes.txt:118]: `versionchanged` annotation is missing.

File: django/contrib/contenttypes/models.py (hunk 2 above)
Comment (felixxm) [reply django/contrib/contenttypes/models.py:185]: Sorry I wanted to add this comment to the second branch.

I don't see any test failures when simplifying `get_object_for_this_type()` to the:
```
def get_object_for_this_type(self, using=None, **kwargs):
    return self.model_class()._base_manager.using(using).get(**kwargs)
```

Do we need a special path when `using` is `None`? If yes, can we add test for it.

File: docs/ref/contrib/contenttypes.txt (hunk 3 above)
Comment (bcail) [reply docs/ref/contrib/contenttypes.txt]: I added a `versionadded` annotation for the `using` argument, but I can change it if that's not right.

Comment (felixxm) [reply docs/ref/contrib/contenttypes.txt]: You should use `versionchanged` as it's a change in the `get_object_for_this_type()`.

File: django/contrib/contenttypes/models.py (hunk 2 above)
Comment (felixxm) [reply django/contrib/contenttypes/models.py]: > the base manager won't have a DB set to a different value than the router default, that might get overridden if we pass in `None` to the `using()` call

Can you add a regression test for this scenario? I think we should keep the previous behavior when `using` is not set, i.e. `self.model_class()._base_manager.using(using or self._state.db).get(**kwargs)` :thinking: What do you think?

File: docs/ref/contrib/contenttypes.txt (hunk 3 above)
Comment (bcail) [reply docs/ref/contrib/contenttypes.txt]: I changed it to `versionchanged`.

File: django/contrib/contenttypes/models.py (hunk 2 above)
Comment (bcail) [reply django/contrib/contenttypes/models.py]: I tried change the code to `using(using or self._state.db)`, and I got the following error:
```
ERROR: test_contenttype_in_separate_db (multiple_database.tests.QueryTestCase.test_contenttype_in_separate_db)
multiple_database.models.Book.DoesNotExist: Book matching query does not exist.
```

I'm not really sure how to add a regression test for this scenario. Does django support having a `_base_manager` pointing to a different database than what the database router points to for those objects? Seems like some parts of django might break in that case.

But I'm not clear on all the details here. I'll look at it more, but any pointers are welcome.

Comment (felixxm) [reply django/contrib/contenttypes/models.py]: > I tried change the code to using(using or self._state.db), and I got the following error:

`book_type` in on the `default` and book is on the `other`, so `get_object_for_this_type()` doesn't return an object, so it's better to pass `None`. I'd add two more assertions for:
- `get_object_for_this_type(using="default")`
- `get_object_for_this_type(using="other")`
to test all possible options.

File: docs/ref/contrib/contenttypes.txt [hunk 4]
```diff
@@ -106,14 +106,19 @@ methods that allow you to get from a
...
     Takes a set of valid :ref:`lookup arguments <field-lookups-intro>` for the
     model the :class:`~django.contrib.contenttypes.models.ContentType`
     represents, and does
     :meth:`a get() lookup <django.db.models.query.QuerySet.get>`
     on that model, returning the corresponding object.
 
+    .. versionchanged:: 5.1
+
+        The ``using`` argument was added, for specifying a different database
+        than the default.
```
Comment (felixxm) [inline docs/ref/contrib/contenttypes.txt:120]:
```suggestion
    .. versionchanged:: 5.1

        The ``using`` argument was added.
```

We can add an explanation how it works to the main `get_object_for_this_type()` docs.

Comment (bcail) [reply docs/ref/contrib/contenttypes.txt]: I updated the docs.

File: django/contrib/contenttypes/models.py (hunk 2 above)
Comment (bcail) [reply django/contrib/contenttypes/models.py]: I added a couple asserts... is that what you had in mind?

File: tests/multiple_database/tests.py [hunk 5]
```diff
@@ -1302,6 +1302,31 @@ def test_related_manager(self):
...
+        self.assertEqual(book.published, book_published_date)
+
+        book = book_type.get_object_for_this_type(using="other", title=book_title)
+        self.assertEqual(book.published, book_published_date)
+
+        with self.assertRaises(Book.DoesNotExist):
+            book_type.get_object_for_this_type(using="default", title=book_title)
+
+        all_books = book_type.get_all_objects_for_this_type()
+        self.assertEqual(len(all_books), 1)
```
Comment (felixxm) [inline tests/multiple_database/tests.py:1328]: I'd add a second book in the `default` db to confirm how it behaves when books are in two different databases. I'd expect that something like this to work:
```suggestion
        ContentType.objects.using("other").all().delete()
        
        book_other = Book.objects.using("other").create(
            title="Test title other", published=datetime.date(2009, 5, 4)
        )
        book_default = Book.objects.using("other").create(
            title="Test title default", published=datetime.date(2009, 5, 4)
        )
        book_type = ContentType.objects.using("default").get(
            app_label="multiple_database", model="book"
        )

        book = book_type.get_object_for_this_type(title=book_other.title)
        self.assertEqual(book, book_other)
        book = book_type.get_object_for_this_type(using="other", title=book_other.title)
        self.assertEqual(book, book_other)
        
        book = book_type.get_object_for_this_type(title=book_default.title)
        self.assertEqual(book, book_default)
        book = book_type.get_object_for_this_type(using="default", title=book_default.title)
        self.assertEqual(book, book_default)

        all_books = book_type.get_all_objects_for_this_type()
        self.assertCountEqual(all_books, [book_default, book_other])
```

Comment (felixxm) [reply tests/multiple_database/tests.py:1328]: Moreover, ticket-25564 and ticket-30760 were marked as a duplicates of this ticket. Can you add a regression test for them to confirm that the "View on site" is also fixed?

Comment (felixxm) [reply tests/multiple_database/tests.py:1328]: It looks like ticket-15610 may be another duplicate, Can you confirm this and add an appropriate regression test?

Comment (bcail) [reply tests/multiple_database/tests.py:1328]: @felixxm I tried your additional tests, but they don't work currently. `book_type.get_object_for_this_type(title=book_default.title)` isn't able to find the book in the default database, and `book_type.get_all_objects_for_this_type()` doesn't check through both databases to find all the objects.

Is it expected that these queries will check across all databases to find objects? Seems like we could add code to fall back to other databases for `get_object_for_this_type()`, and we could query all databases in `get_all_objects_for_this_type()`, but would we run into issues with duplicate primary keys if we're adding together lists of objects from multiple databases?

Comment (bcail) [reply tests/multiple_database/tests.py:1329]: @felixxm I added a test for View on Site.

For ticket 15610, I added a comment with a question on that issue.

Comment (felixxm) [reply tests/multiple_database/tests.py:1329]: > @felixxm I tried your additional tests, but they don't work currently. `book_type.get_object_for_this_type(title=book_default.title)` isn't able to find the book in the default database, and `book_type.get_all_objects_for_this_type()` doesn't check through both databases to find all the objects.
> 
> Is it expected that these queries will check across all databases to find objects? Seems like we could add code to fall back to other databases for `get_object_for_this_type()`, and we could query all databases in `get_all_objects_for_this_type()`, but would we run into issues with duplicate primary keys if we're adding together lists of objects from multiple databases?

Agreed we cannot check all databases, so it should work as follows:
```python
    @override_settings(DATABASE_ROUTERS=["multiple_database.tests.TestRouter"])
    def test_contenttype_in_separate_db(self):
        book_other = Book.objects.using("other").create(
            title="Test title other", published=datetime.date(2009, 5, 4)
        )
        book_default = Book.objects.using("default").create(
            title="Test title default", published=datetime.date(2009, 5, 4)
        )
        book_type = ContentType.objects.using("default").get(
            app_label="multiple_database", model="book"
        )
        book = book_type.get_object_for_this_type(title=book_other.title)
        self.assertEqual(book, book_other)
        book = book_type.get_object_for_this_type(using="other", title=book_other.title)
        self.assertEqual(book, book_other)
        
        with self.assertRaises(Book.DoesNotExist):
            book = book_type.get_object_for_this_type(title=book_default.title)
        book = book_type.get_object_for_this_type(using="default", title=book_default.title)
        self.assertEqual(book, book_default)
        
        all_books = book_type.get_all_objects_for_this_type()
        self.assertCountEqual(all_books, [book_other])

```

**Comment by felixxm** [comment]
@tfrydrychewicz Does it work for you?

File: tests/multiple_database/tests.py [hunk 6]
```diff
@@ -1302,6 +1303,61 @@ def test_related_manager(self):
...
+        ContentType.objects.using("other").all().delete()
+        book_title = "Test Content Type in Separate DB Admin View on Site"
+        book_published_date = datetime.date(2009, 5, 4)
+        book = Book.objects.using("other").create(
+            title=book_title, published=book_published_date
+        )
+        book_id = book.id
+
+        with override_settings(
+            DATABASE_ROUTERS=["multiple_database.routers.AdminRouter"],
```
[the remark this answers is not in the document]
Comment (bcail) [reply tests/multiple_database/tests.py:1341]: Thanks - I updated the router.

File: tests/multiple_database/tests.py (hunk 5 above)
Comment (bcail) [reply tests/multiple_database/tests.py]: OK, I updated that test - hopefully it looks OK now.

File: tests/multiple_database/tests.py [hunk 7]
```diff
@@ -1302,6 +1303,61 @@ def test_related_manager(self):
...
+            )
+
+            self.client.force_login(user)
+
+            shortcut_url = reverse("admin:view_on_site", args=(book_type.pk, book_id))
+            response = self.client.get(shortcut_url, follow=False)
+            self.assertEqual(response.status_code, 302)
+            self.assertRegex(
+                response.url, f"http://(testserver|example.com)/books/{book_id}/"
+            )
```
[the remark this answers is not in the document]
Comment (bcail) [reply tests/multiple_database/tests.py]: I moved the test.

**Comment by felixxm** [comment]
@bcail Thanks :+1: 

Merged in 02a600ff67f7b106cdcab22310bacea98c1a26ba.

**Pull request closed without merging**